Add guide on Rancher impersonation (#873)
rancher/rancher-docs: docs/reference-guides/rancher-manager-architecture/communicating-with-downstream-user-clusters.md

The authentication proxy forwards all Kubernetes API calls to downstream clusters. It integrates with authentication services like local authentication, Active Directory, and GitHub. On every Kubernetes API call, the authentication proxy authenticates the caller and sets the proper Kubernetes impersonation headers before forwarding the call to Kubernetes masters.

> Rancher communicates with Kubernetes clusters using a [service account,](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/). Each user in Rancher correlates with their own service account in the downstream cluster. Rancher uses the service account to [impersonate](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) the user, in order to assume all the permissions the user is meant to have.

Author: :+1:

You will need to use a context defined in this kubeconfig file to access the cluster if Rancher goes down. Therefore, we recommend exporting the kubeconfig file so that if Rancher goes down, you can still use the credentials in the file to access your cluster. For more information, refer to the section on accessing your cluster with [kubectl and the kubeconfig file.](../../how-to-guides/new-user-guides/manage-clusters/access-clusters/use-kubectl-and-kubeconfig.md)

## Impersonation

> Users exist only on the Rancher cluster. When they interact with a downstream cluster through the authentication proxy, there needs to be some entity on the cluster to be the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission, which is to **impersonate** the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).

Author: :+1:

> Users technically exist only on the upstream cluster. When users interact with a downstream cluster through the authentication proxy, there needs to be some entity downstream to serve as the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission, which is to **impersonate** the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).

Reviewer: > If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).

I feel like this part of the text isn't necessary. This paragraph seems more clear to me without it, not sure if you agree though. Do you feel the need to explain the design decision wrt. impersonation?

Reviewer 2: I like this section just because it was my first question after reading the above statement.

### Impersonation Troubleshooting

> Five resources are created to handle impersonation:

Author: :+1:

> Five resources handle impersonation:

Reviewer: I think it's worth mentioning (again) where those resources are located. The previous section mentions both upstream and downstream cluster so at this point I'm unsure where these resources are located
(eg: users only exist on upstream cluster, service account on downstream, so those would be on... downstream, I think)

Author: :+1:

- namespace: `cattle-impersonation-system`
- service account: `cattle-impersonation-system/cattle-impersonation-<user ID>`
- account token secret: `cattle-impersonation-system/cattle-impersonation-<user ID>-token-<hash>`
- cluster role: `cattle-impersonation-<user ID>`
- cluster role binding: `cattle-impersonation-<user ID>`

> Below is an example of what the cluster role usually looks like. In this example, the system is configured to use `github` as the auth provider:

Author: :+1:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
 creationTimestamp: "2021-10-06T18:20:13Z"
 labels:
   authz.cluster.cattle.io/impersonator: "true"
   cattle.io/creator: norman
 name: cattle-impersonation-user-abcde
 resourceVersion: "3528"
 uid: a7478731-72a0-4343-b09f-c3bf12552d77
rules:
- apiGroups: # allowed to impersonate user user-abcde
 - ""
 resourceNames:
 - user-abcde
 resources:
 - users
 verbs:
 - impersonate

> - apiGroups: # allowed to impersonate groups github_team://123 (group from GitHub auth provider), system:authenticated, system:cattle:authenticated (automatic groups from Rancher and Kubernetes)
>  - ""

Reviewer: I feel like the following would be easier to read because the codeblock clips the text so you need to scroll left/right to view the rest of it. Similarly for the next apigroups. 

```suggestion
# allowed to impersonate the following groups:
# - github_team://123 (group from GitHub auth provider)
# - system:authenticated
# - system:cattle:authenticated (automatic groups from Rancher and Kubernetes)
- apiGroups:
  - ""
```

What do you think?

Author: Done, but added comments to the resource names section instead of adding another bulleted list

> - apiGroups: # allowed to impersonate groups github_team://123 (group from github auth provider), system:authenticated, system:cattle:authenticated (automatic groups from rancher and k8s)

Reviewer 3: ```suggestion
- apiGroups: # allowed to impersonate groups github_team://123 (group from Github auth provider), system:authenticated, system:cattle:authenticated (automatic groups from Rancher and Kubernetes)
```

Author: :+1:

 resourceNames:
 - github_team://123
 - system:authenticated
 - system:cattle:authenticated
 resources:
 - groups
 verbs:
 - impersonate

> - apiGroups: # allowed to impersonate principal ID github_user://098 (principal ID from github auth provider)

Reviewer 3: ```suggestion
- apiGroups: # allowed to impersonate principal ID github_user://098 (principal ID from Github auth provider)
```

Author: :+1:

 - authentication.k8s.io
 resourceNames:
 - github_user://098
 resources:
 - userextras/principalid
 verbs:
 - impersonate

> - apiGroups: # allowed to impersonate username example (username from github auth provider)

Reviewer 3: ```suggestion
- apiGroups: # allowed to impersonate username example (username from Github auth provider)
```

Author: :+1:

 - authentication.k8s.io
 resourceNames:
 - example
 resources:
 - userextras/username
 verbs:
 - impersonate
```

> When you troubleshoot impersonation issues, check whether these resources exist for the user, and check whether the rules in the cluster role look similar to the above.

Author: :+1:

> When troubleshooting impersonation issues, check whether these resources exist for the user, and check whether the rules in the cluster role look similar to the above.

Reviewer 3: ```suggestion
When you troubleshoot impersonation issues, check whether these resources exist for the user, and check whether the rules in the cluster role look similar to the above.
```

Author: :+1:

> If you see an error related to "impersonation" in the UI, pay close attention to the **end** of the error message, which should indicate the real reason the request failed.

Reviewer 3: ```suggestion
If you see an error related to "impersonation" in the UI, pay close attention to the **end** of the error message, which should indicate the real reason that the request failed.
```

Author: :+1:

You will need to use a context defined in this kubeconfig file to access the cluster if Rancher goes down. Therefore, we recommend exporting the kubeconfig file so that if Rancher goes down, you can still use the credentials in the file to access your cluster. For more information, refer to the section on accessing your cluster with [kubectl and the kubeconfig file.](../../how-to-guides/new-user-guides/manage-clusters/access-clusters/use-kubectl-and-kubeconfig.md)

## Impersonation

> Users technically exist only on the upstream cluster. When users interact with a downstream cluster through the authentication proxy, there needs to be some entity downstream to serve as the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission, which is to **impersonate** the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).

Reviewer 2: We should add some wording to specify how we determine downstream permissions. At first, I was confused about how the service account can perform action if it only has impersonation rules.
```suggestion
Users technically exist only on the upstream cluster. Eventhough all RoleBindings and ClusterRoleBindings on the downstream refer to the userInfo stored in the local cluster. When users interact with a downstream cluster through the authentication proxy, there needs to be some entity downstream to serve as the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission, which is to **impersonate** the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).
```

Reviewer: I'm not sure I understand. Doesn't the CR/CRBs in downstream refer to the `user-*` created in the downstream?

Reviewer: Alright, less confused now.

> Eventhough all RoleBindings and ClusterRoleBindings on the downstream refer to the userInfo stored in the local cluster. 

I don't think userInfo is mentioned anywhere in the Rancher docs so that might confuse people that don't know about it.

Author: Added some references to upstream RBAC docs for clarity

> Users technically exist only on the upstream cluster. Rancher creates [RoleBindings and ClusterRoleBindings](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding) that refer to Rancher users, even though there is [no actual User resource](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes) on the downstream cluster. When users interact with a downstream cluster through the authentication proxy, there needs to be some entity downstream to serve as the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission, which is to **impersonate** the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).

Reviewer 3: ```suggestion
Users technically exist only on the upstream cluster. Rancher creates [RoleBindings and ClusterRoleBindings](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding) that refer to Rancher users, even though there is [no actual User resource](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes) on a downstream cluster. 

When users interact with a downstream cluster through the authentication proxy, there needs to be some entity downstream to serve as the actor for those requests. Rancher creates service accounts to be that entity. Each service account is only granted one permission; namely, permission to *impersonate* the user they belong to. If there was only one service account that could impersonate any user, then it would be possible for a malicious user to corrupt that account and escalate their privileges by impersonating another user. This issue was the basis for a [CVE](https://github.com/rancher/rancher/security/advisories/GHSA-pvxj-25m6-7vqr).
```

### Impersonation Troubleshooting

On the downstream cluster, five resources handle impersonation:

- namespace: `cattle-impersonation-system`
- service account: `cattle-impersonation-system/cattle-impersonation-<user ID>`
- account token secret: `cattle-impersonation-system/cattle-impersonation-<user ID>-token-<hash>`
- cluster role: `cattle-impersonation-<user ID>`
- cluster role binding: `cattle-impersonation-<user ID>`

In this example of a typical impersonation cluster role, the system is configured to use `github` as the auth provider:

> ```

Reviewer: ```suggestion
```yaml
```

I missed that on the first review sorry.  I can see that the highlighting is set to yaml in a lot of other pages so I think it would make sense to have it here as well.

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
 creationTimestamp: "2021-10-06T18:20:13Z"
 labels:
   authz.cluster.cattle.io/impersonator: "true"
   cattle.io/creator: norman
 name: cattle-impersonation-user-abcde
 resourceVersion: "3528"
 uid: a7478731-72a0-4343-b09f-c3bf12552d77
rules:
# allowed to impersonate user user-abcde
- apiGroups:
 - ""
 resourceNames:
 - user-abcde
 resources:
 - users
 verbs:
 - impersonate
# allowed to impersonate listed groups
- apiGroups:
 - ""
 resourceNames:
 - github_team://123 # group from GitHub auth provider
 - system:authenticated # automatic group from Kubernetes
 - system:cattle:authenticated # automatic group from Rancher
 resources:
 - groups
 verbs:
 - impersonate
# allowed to impersonate principal ID github_user://098
- apiGroups:
 - authentication.k8s.io
 resourceNames:
 - github_user://098 # principal ID from GitHub auth provider
 resources:
 - userextras/principalid
 verbs:
 - impersonate
# allowed to impersonate username example
- apiGroups:
 - authentication.k8s.io
 resourceNames:
 - example # username from GitHub auth provider
 resources:
 - userextras/username
 verbs:
 - impersonate
```

> When you troubleshoot impersonation issues, check whether these resources exist for the user, and check whether the rules in the cluster role look similar to the above. For example:

Reviewer 3: ```suggestion
When you troubleshoot impersonation issues, check whether these resources exist for the user, and whether the rules in the cluster role look similar to the above. For example:
```

```bash
kubectl --namespace cattle-impersonation-system get serviceaccount cattle-impersonation-<user ID>
kubectl --namespace cattle-impersonation-system get secret cattle-impersonation-<user ID>-token-<hash>
kubectl get clusterrole cattle-impersonation-<user ID> --output yaml
kubectl get clusterrolebinding cattle-impersonation-<user ID>
```

> If you see an error related to "impersonation" in the UI, pay close attention to the **end** of the error message, which should indicate the real reason that the request failed.

Reviewer 3: ```suggestion
If you see an error related to "impersonation" in the UI, pay close attention to the *end* of the error message, which should indicate the real reason that the request failed.
```
